Sources/Purchasing/IntroEligibilityCalculator.swift

```diff
@@ -86,26 +90,36 @@ extension IntroEligibilityCalculator: @unchecked Sendable {}
 @available(iOS 12.0, macOS 10.14, macCatalyst 13.0, tvOS 12.0, watchOS 6.2, *)
 private extension IntroEligibilityCalculator {
 
-    func checkEligibility(candidateProducts: Set<StoreProduct>,
-                          purchasedProductsWithIntroOffers: Set<StoreProduct>) -> [String: IntroEligibilityStatus] {
+    func checkEligibility(
+        candidateProducts: Set<StoreProduct>,
+        purchasedProductsWithIntroOffersOrFreeTrials: Set<StoreProduct>,
+        activeSubscriptionProducts: Set<StoreProduct>
+    ) -> [String: IntroEligibilityStatus] {
         var result: [String: IntroEligibilityStatus] = [:]
 
         for candidate in candidateProducts {
             guard candidate.subscriptionPeriod != nil else {
                 result[candidate.productIdentifier] = IntroEligibilityStatus.unknown
                 continue
             }
-            let usedIntroForProductIdentifier = purchasedProductsWithIntroOffers
-                .contains { purchased in
-                    let foundByGroupId = (candidate.subscriptionGroupIdentifier != nil
-                        && candidate.subscriptionGroupIdentifier == purchased.subscriptionGroupIdentifier)
-                    return foundByGroupId
+            let usedIntroForProductIdentifier = (
+                candidate.subscriptionGroupIdentifier != nil &&
+                purchasedProductsWithIntroOffersOrFreeTrials.contains { purchased in
+                    candidate.subscriptionGroupIdentifier == purchased.subscriptionGroupIdentifier
+                }
+            )
+            let usedIntroInSubscriptionGroup = (
+                candidate.subscriptionGroupIdentifier != nil &&
+                activeSubscriptionProducts.contains { purchased in
+                    candidate.subscriptionGroupIdentifier == purchased.subscriptionGroupIdentifier
                 }
+            )
```

Comment:
aren't we missing the ones with subsGroupID == nil?

Reply:
Maybe. Can we go through this together next week? My brain might have been fried last week, but couldn't figure out how to make all the tests pass.

Reply:
absolutely!